runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java

```diff
@@ -68,83 +71,107 @@ public ILogData peek(RuntimeLayout runtimeLayout, long globalAddress) {
         log.trace("Read[{}]: chain {}/{}", globalAddress, numUnits, numUnits);
         // In chain replication, we read from the last unit, though we can optimize if we
         // know where the committed tail is.
-        ILogData ret = CFUtils.getUninterruptibly(
-                runtimeLayout
-                        .getLogUnitClient(globalAddress, numUnits - 1)
-                        .read(globalAddress)).getAddresses()
-                .getOrDefault(globalAddress, null);
-        return ret == null || ret.isEmpty() ? null : ret;
+        ILogData peekResult = CFUtils.getUninterruptibly(runtimeLayout
+                .getLogUnitClient(globalAddress, numUnits - 1)
+                .read(globalAddress)).getAddresses().get(globalAddress);
+
+        return peekResult.isEmpty() ? null : peekResult;
     }
 
     /**
-     * {@inheritDoc}
+     * Reads a list of global addresses from the chain of log unit servers.
+     *
+     * - This method optimizes for the time to wait to hole fill in case an empty address
+     * is encountered.
+     * - If the waitForWrite flag is set to true, when an empty address is encountered,
+     * it waits for the hole to be filled. All subsequent empty addresses are hole filled
+     * directly and the reader does not wait.
+     * - In case the flag is set to false, none of the reads wait for write completion and
+     * the empty addresses are hole filled right away.
+     *
+     * @param runtimeLayout runtime layout.
+     * @param addresses     list of addresses to read.
+     * @param waitForWrite  flag whether wait for write is required or hole fill directly.
+     * @return Map of read addresses.
      */
     @Override
-    public Map<Long, ILogData> readAll(RuntimeLayout runtimeLayout, List<Long> globalAddresses) {
-        Range<Long> range = Range.encloseAll(globalAddresses);
-        return readRange(runtimeLayout, range, true);
-    }
+    @Nonnull
+    public Map<Long, ILogData> readAll(RuntimeLayout runtimeLayout,
+                                       List<Long> addresses,
+                                       boolean waitForWrite) {
+
+        // A map of log unit server endpoint to addresses it's responsible for
+        Map<String, List<Long>> serverAddressMap = new HashMap<>();
 
+        for (Long address : addresses) {
+            List<String> logServers = runtimeLayout.getLayout().getStripe(address).getLogServers();
```

Comment:
If you can, I suggest you also use `public LayoutStripe getStripe(long globalAddress)` to implement the "segment finding logic" in `getStripe`

Reply:
Done. Cleaned up getStripe() method.